presto-main/src/main/java/com/facebook/presto/operator/PagesRTreeIndex.java

```diff
@@ -0,0 +1,190 @@
+            Session session,
+            LongArrayList addresses,
+            List<Type> types,
+            List<Integer> outputChannels,
+            List<List<Block>> channels,
+            STRtree rtree,
+            BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest,
+            Optional<JoinFilterFunctionFactory> filterFunctionFactory)
+    {
+        this.types = types;
+        this.outputChannels = outputChannels;
+        this.channels = requireNonNull(channels, "channels is null");
+        this.rtree = requireNonNull(rtree, "rtree is null");
+        this.spatialRelationshipTest = requireNonNull(spatialRelationshipTest, "spatial relationship is null");
+        this.filterFunction = filterFunctionFactory.map(factory -> factory.create(session.toConnectorSession(), addresses, channels));
+        this.memorySizeInBytes = computeMemorySizeInBytes(addresses);
+    }
+
+    private static Envelope getEnvelope(OGCGeometry ogcGeometry)
+    {
+        com.esri.core.geometry.Envelope env = new com.esri.core.geometry.Envelope();
+        ogcGeometry.getEsriGeometry().queryEnvelope(env);
+
+        return new Envelope(env.getXMin(), env.getXMax(), env.getYMin(), env.getYMax());
+    }
+
+    private long computeMemorySizeInBytes(LongArrayList addresses)
+    {
+        return INSTANCE_SIZE + STRTREE_INSTANCE_SIZE + computeMemorySizeInBytes(rtree.getRoot()) +
+                sizeOf(addresses.elements()) + channels.stream()
+                    .flatMap(List::stream)
+                    .mapToLong(Block::getRetainedSizeInBytes)
+                    .sum();
+    }
+
+    private long computeMemorySizeInBytes(AbstractNode root)
+    {
+        if (root.getLevel() == 0) {
+            return ABSTRACT_NODE_INSTANCE_SIZE + ENVELOPE_INSTANCE_SIZE + root.getChildBoundables().stream().mapToLong(child -> computeMemorySizeInBytes((ItemBoundable) child)).sum();
+        }
+        else {
+            return ABSTRACT_NODE_INSTANCE_SIZE + ENVELOPE_INSTANCE_SIZE + root.getChildBoundables().stream().mapToLong(child -> computeMemorySizeInBytes((AbstractNode) child)).sum();
+        }
+    }
+
+    private long computeMemorySizeInBytes(ItemBoundable item)
+    {
+        return ENVELOPE_INSTANCE_SIZE + ((GeometryWithAddress) item.getItem()).getEstimatedMemorySizeInBytes();
+    }
+
+    @Override
+    public long getEstimatedMemorySizeInBytes()
+    {
+        return memorySizeInBytes;
+    }
+
+    // Returns an array of addresses from PagesIndex.valueAddresses
+    @Override
+    public long[] findJoinAddresses(int probePosition, Page probe, int probeGeometryChannel)
+    {
+        Block probeGeometryBlock = probe.getBlock(probeGeometryChannel);
+        if (probeGeometryBlock.isNull(probePosition)) {
+            return EMPTY_ADDRESSES;
+        }
+
+        Slice slice = probeGeometryBlock.getSlice(probePosition, 0, probeGeometryBlock.getSliceLength(probePosition));
+        OGCGeometry probeGeometry = deserialize(slice);
```

Comment:
There's a lot of (unnecessary?) coupling between this class and the properties of the data in the input streams (the fact that they are deserializable as OGCGeometry, etc). It'd be ideal if this class were:
1. agnostic about the fact it's dealing with geometry objects from ESRI
2. able to delegate to functions provided by the plugin (bound in at the time the operator is created). This method, in particular, seems to be doing some stuff that functions like st_contains, etc already does.